### DISCUSSION TITLE
Reflection with code generation

### DISCUSSION THREAD

**Discussion by h4sjohnson** [nxrighthere/UnrealCLR#145] [Ideas]
I‘ve just finished reading and comprehending your amazing work. Thanks for bring C# as an Unreal alternative progamming language.
I come up an idea that using UE reflection system to grab the information(UFunction, UClass...) exposed to blueprint, then generate C# code from that information. I've noticed a file named "codegen.cs", but we can definitely do much more with code generation. Some examples on "FPyWrapperTypeRegistry::GenerateWrappedTypes" and "FPyWrapperTypeRegistry::GenerateStubCodeForWrappedTypes". These code wrap C++ to python.
I am currently working on proofing this idea.

**Comment by h4sjohnson** [comment]
By the way, why Object is not the super class of ActorComponent?

**Reply by nxrighthere** [reply]
Some of the `Object` API might be misleading and confusing for the end-user. I organized and implemented it where appropriate instead of using inheritance.

**Comment by nxrighthere** [comment]
> I come up an idea that using UE reflection system to grab the information(UFunction, UClass...) exposed to blueprint, then generate C# code from that information. I've noticed a file named "codegen.cs", but we can definitely do much more with code generation. Some examples on "FPyWrapperTypeRegistry::GenerateWrappedTypes" and "FPyWrapperTypeRegistry::GenerateStubCodeForWrappedTypes". These code wrap C++ to python.

I have a few test implementations based on the reflection similar to the Python plugin, but there are several caveats that I'm trying to solve for quite a while.

> I‘ve just finished reading and comprehending your amazing work. Thanks for bring C# as an Unreal alternative progamming language.

Thank you.
